lucene/spatial-extras/src/java/org/apache/lucene/spatial/prefix/tree/S2PrefixTreeCell.java

```diff
@@ -0,0 +1,285 @@
+ */
+class S2PrefixTreeCell implements Cell {
+
+    //Faces of S2 Geometry
+    private static S2CellId[] FACES = new S2CellId[6];
+    static {
+        FACES[0] = S2CellId.fromFacePosLevel(0, 0, 0);
+        FACES[1] = S2CellId.fromFacePosLevel(1, 0, 0);
+        FACES[2] = S2CellId.fromFacePosLevel(2, 0, 0);
+        FACES[3] = S2CellId.fromFacePosLevel(3, 0, 0);
+        FACES[4] = S2CellId.fromFacePosLevel(4, 0, 0);
+        FACES[5] = S2CellId.fromFacePosLevel(5, 0, 0);
+    }
+
+    /*Special character to define a cell leaf*/
+    private static final byte LEAF = '+';
+
+    /*Tokens are used to serialize cells*/
+    private static final byte[] TOKENS;
+    /*Map containing mapping between tokens and integer values*/
+    private static final Map<Byte,Integer> PIXELS;
+    static {
+        TOKENS = new byte[]{'0', '1', '2', '3', '4', '5'};
+        PIXELS = new HashMap<>(6);
+        PIXELS.put(TOKENS[0], 0);
+        PIXELS.put(TOKENS[1], 1);
+        PIXELS.put(TOKENS[2], 2);
+        PIXELS.put(TOKENS[3], 3);
+        PIXELS.put(TOKENS[4], 4);
+        PIXELS.put(TOKENS[5], 5);
+    }
+
+    S2CellId cellId;
+    int level; //cache level
+    S2PrefixTree tree;
+
+    SpatialRelation shapeRel= null;
+    boolean isLeaf;
+    Shape shape = null;
+
+    S2PrefixTreeCell(S2PrefixTree tree, S2CellId cellId){
+        this.cellId= cellId;
+        this.tree = tree;
+        setLevel();
+        if (getLevel() == tree.getMaxLevels()) {
+            setLeaf();
+        }
+    }
+
+    void readCell(S2PrefixTree tree, BytesRef ref){
+        isLeaf = false;
+        shape = null;
+        shapeRel = null;
+        this.tree = tree;
+        cellId = getS2CellIdFromBytesRef(ref);
+        setLevel();
+        if (isLeaf(ref) || getLevel() == tree.getMaxLevels()){
+            setLeaf();
+        }
+    }
+
+    @Override
+    public SpatialRelation getShapeRel() {
+        return shapeRel;
+    }
+
+    @Override
+    public void setShapeRel(SpatialRelation rel) {
+        shapeRel = rel;
+    }
+
+    @Override
+    public boolean isLeaf() {
+        return isLeaf;
+    }
+
+    @Override
+    public void setLeaf() {
+        isLeaf = true;
+    }
+
+    @Override
+    public BytesRef getTokenBytesWithLeaf(BytesRef result) {
+        result = getTokenBytesNoLeaf(result);
+        //max levels do not have leaf
+        if (isLeaf() && !(getLevel() == tree.getMaxLevels())){
+            //Add leaf byte
+            result.bytes[result.offset + result.length] = LEAF;
+            result.length++;
+        }
+        return result;
+    }
+
+    @Override
+    public BytesRef getTokenBytesNoLeaf(BytesRef result) {
+        if (result == null){
+            result = new BytesRef();
+        }
+        getBytesRefFromS2CellId(cellId, result);
+        return result;
+    }
+
+    @Override
+    public int getLevel() {
+        return this.level;
+    }
+
+    /**
+     * Cache level of cell.
+     */
+    private void setLevel() {
```

Comment:
The method cellId.level() is probably efficient but because it can be called several times I think we are saving some ticks on the clock by caching it.